rust/template/tcp_channel/lib.rs

```diff
@@ -0,0 +1,262 @@
+        spawn(move || {
+            let (socket, _) = listener.accept()?;
+            *stream.lock().unwrap() = Some(socket.as_raw_fd());
+
+            let reader = BufReader::new(socket);
+            let mut lines = reader.lines();
+            loop {
+                let mut upds = lines
+                    .by_ref()
+                    .take_while(|result| match result {
+                        Ok(line) if line == "commit" => false,
+                        Ok(_) => true,
+                        Err(_) => false,
+                    })
+                    .map(|result| {
+                        result.map(|line| {
+                            let v: T = from_str(&line).unwrap();
+                            v
+                        })
+                    });
+
+                let first = match upds.next() {
+                    None => break,
+                    Some(v @ Ok(_)) => v,
+                    Some(Err(e)) => return Err(e),
+                };
+
+                // If there is no observer we just drop the data, which
+                // is seemingly the only reasonable behavior given that
+                // observers can come and go by virtue of our API
+                // design.
+                if let Some(ref mut observer) = observer.lock().unwrap().deref_mut() {
+                    let upds = Some(first).into_iter().chain(upds).map(Result::unwrap);
+                    // TODO: Need to handle those errors eventually (or
+                    //       perhaps we will end up with method
+                    //       signatures that don't allow for errors?).
+                    observer.on_start().unwrap();
+                    observer.on_updates(Box::new(upds)).unwrap();
+                    observer.on_commit().unwrap();
+                }
+            }
+            Ok(())
+        })
+    }
+
+    /// Retrieve the address we are listening on.
+    pub fn addr(&self) -> &SocketAddr {
+        &self.addr
+    }
+}
+
+impl<T> Drop for TcpReceiver<T> {
+    fn drop(&mut self) {
+        // We can't really handle any errors here.
+        // The order of operations is important. First close the
+        // listener to be sure that a no new stream will be accepted,
+        // then close the stream itself.
+        unsafe {
+            let _ = close(self.listener);
+            let _ = self.stream.lock().unwrap().map(|fd| close(fd));
+        }
+        // The remaining members will be destroyed automatically, no
+        // need to bother here.
+    }
+}
+
+impl<T> Observable<T, String> for TcpReceiver<T>
+where
+    T: Debug + Send + 'static,
+{
+    /// An observer subscribes to the receiving end of a TCP channel to
+    /// listen to incoming data.
+    fn subscribe(
+        &mut self,
+        observer: Box<dyn Observer<T, String> + Send>,
+    ) -> Box<dyn Subscription> {
+        let mut guard = self.observer.lock().unwrap();
+        match *guard {
+            Some(_) => panic!(
```

Comment:
This really should not panic. Option 1: quietly replace the existing observer; Option 2: return an error; Option 3: replace the existing observer and return the old one. Options 1 and 3 do not really work in the presence of the `Subscription` atrocity. I think Option 2 is probably the simplest one, but requires changing the API to allow errors. Whatever semantics we choose must be part of the contract for all observables.

Thoughts?

Reply:
I'll make it return an error. This seems like the sanest approach to me, from a usability point of view.